### ISSUE TITLE
ImVector size and capacity stored as int?

### ISSUE THREAD

**Issue by Ybalrid** [ocornut/imgui#1661]
Hi,

I was looking into the code of the ImVector class, and I'm a bit surprised to find that the size/capacity of the vector are stored in `int` variables.

I would expect this information to be stored in `size_t` variables, as it is by definition the type to use for referring to the size of a container. (and the index inside a sequence container with random access). It's unsigned and is guaranteed to be large enough for size size of the biggest vector you can have. 

Also, even if it's not stated in standard, it seems that the usage of size_t in this scenarist (especially when iterating through containers) is *slightly* better optimized by compilers. (e.g. see this : https://twitter.com/lefticus/status/940608074426941445)

Would you accept to merge a PR that replaces the usage of `int` with `size_t` in this scenario? If so I'm happy to submit one :wink:

**Comment by ocornut** [comment]
Hello,

There's some truth to that but the tweet and code relying std::vector are also partially misleading. "Correctness" is ill-defined. The specs of ImVector is that it can't index over 2 billion entries, so be it? How about the indirect effect on cache/memory of making many of your structures larger? I adopted int a long time ago because size_t-correctness leads to recurrent/endless 32/64 casting and warnings in both imgui and any user code base, I deemed it was simpler to stick to int and it's was very liberating change. Especially as one of the thing we are combating here is complexity.

I agree perf wise it could be better but I also suspect it's barely measurable for our use case.
This is a good read if you are interested in the topic:
https://gist.github.com/rygorous/e0f055bfb74e3d5f0af20690759de5a7

Ideally I believe we should store u32, use size_t for loop iteration, and use ~0 instead of -1 if we need a special marker.

> guaranteed to be large enough for size size of the biggest vector you can have.

This doesn't matter for realistic usage of imgui.

> Would you accept to merge a PR t

At the moment I would rather not, but I appreciate the thought - the extra breakage/warnings in user's code are not worth it. May reconsider it for imgui 2.0.

**Comment by ice1000** [comment]
I've tried this -- in fact, I've just reverted my 20 commits about changing `int` to `size_t`. Once you change the `Size` member and the `size()` method's return type to `size_t`, you'll see tons of warnings about comparisons between signed values and unsigned values in those for-loops (there's a lot of `for (int n = 0; n < SomeImVectorHere.size(); ...` which is extremely hard to change all of them).

Sometimes there're tricks rely on negative values, which are not compatible with `size_t`s.

But this is just one reason for me to give up `size_t`s -- the `ImTextStr**` functions return `int` as the number of converted utf8 characters. This should also theoretically be `size_t`s instead of `int`s, but it relies on the return value of `vsnprintf`, which is possible to be negative. This is the other reason.

I'm not trying to persuade anyone to use `int` or `size_t` -- I prefer `size_t` too, but it's just very hard work and I've listed two problems that worth worry about. If imgui can handle `int`s and `size_t`s properly internally I'd be very happy.
